core/field_textinput.js

```diff
@@ -278,11 +311,17 @@ Blockly.FieldTextInput.prototype.widgetDispose_ = function() {
   this.isBeingEdited_ = false;
   // No need to call setValue because if the widget is being closed the
   // latest input text has already been validated.
-  if (this.value_ !== this.text_) {
+
+  // This handles the cleaning up of the text input editor. If the display
+  // value is currently different than the actual value, we want to rectify
+  // that now.
+  // As stated above setValue has already been called based on the final
+  // input from the user.
+  if (this.value_ !== this.displayValue_) {
```

Comment:
This is almost the right behavior. Instead we can just set displayValue_ to null as the dispose and handle it in getDisplayText as commented above.